codelists/models.py

```diff
@@ -601,68 +600,28 @@ def table_with_fixed_headers(self, include_mapped_vmps=True):
-            # the codelist
 
             # mapped_vmps_for_this_codelist is a full list of (vmp, previous) tuples, where one of
             # vmp/previous are in the codelist. It may contain mappings where `previous` is more than
             # one historical step away from `vmp`
 
-            # Count if there were any duplicate previous VMPs in the mapping
-            if vmp_to_prev_mapping:
-                counter = Counter(vmp_to_prev_mapping.values())
-                _, count = counter.most_common(1)[0]
-            else:
-                count = 0
-
-            # Since we're mapping later codes as well as previous ones, we also need the
-            # reverse mapping of previous VMP to the one that replaced it
-            prev_to_vmp_mapping = {v: k for k, v in vmp_to_prev_mapping.items()}
-            # If there were any duplicate IMMEDIATE previous VMPs (i.e. two VMPs mapped to the same
-            # previous VMP), we join them in the reversed mapping
-            # This is only needed for the description of where the mapped code comes from
-            if count > 1:
-                duplicate_mappings = [
-                    vmp for vmp, count in counter.items() if count > 1
-                ]
-                for mapped_vmp in duplicate_mappings:
-                    mapped_to = [
-                        vmp
-                        for vmp, prev in vmp_to_prev_mapping.items()
-                        if prev == mapped_vmp
-                    ]
-                    prev_to_vmp_mapping[mapped_vmp] = ", ".join(mapped_to)
-
-            previous_vmps_to_add = set()
-            subsequent_vmps_to_add = set()
-
-            for vmp, previous_vmp in mapped_vmps_for_this_codelist:
+            # create a mapping of previous and subsequent VMPs to add, where the
+            # keys are the mapped VMPs, and the values are the code(s) in the
+            # codelist that they map to/from. Usually there is just a 1:1 mapping, but
+            # it's possible that a single VMP could be split and map to 2 new VMPs, or
+            # that 2 VMPs could be merged to a single new code.
+            previous_vmps_to_add = {}
+            subsequent_vmps_to_add = {}
+
+            for vmp, previous in mapped_vmps_for_this_codelist:
                 if vmp in codes:
-                    # mapping a code in the codelist to a previous code
-                    # add the previous code to the list
-                    previous_vmps_to_add.add(previous_vmp)
-                    # the previous vmp may not be in the prev_to_vmp_mapping mapping
-                    # yet, if it was more than one step away from the code in the codelist.
-                    if previous_vmp not in prev_to_vmp_mapping:
-                        prev_to_vmp_mapping[previous_vmp] = vmp
+                    # mapping in a previous code
+                    previous_vmps_to_add.setdefault(previous, []).append(vmp)
```

Comment:
I'd have reached for `collections.defaultdict` here. I didn't know about `dict.setdefault` 👍🏻